chain-api/src/types/sale.ts

```diff
@@ -0,0 +1,503 @@
+  })
+  @IsNotEmpty()
+  public readonly tokenSaleId: string;
+}
+
+@JSONSchema({
+  description:
+    "Legacy FetchTokenSwapsDto. Provided created timestamp to limit result, or leave empty to query all swaps."
+})
+export class FetchTokenSalesWithPaginationDto extends ChainCallDTO {
+  static readonly MAX_LIMIT = 10 * 1000;
+  static readonly DEFAULT_LIMIT = 1000;
+
+  @JSONSchema({
+    description: "(optional). User alias of the creating user."
+  })
+  @IsOptional()
+  @IsUserAlias()
+  public owner?: string;
+
+  @JSONSchema({
+    description: "Page bookmark. If it is undefined, then the first page is returned."
+  })
+  @IsOptional()
+  @IsNotEmpty()
+  bookmark?: string;
+
+  @JSONSchema({
+    description:
+      `Page size limit. ` +
+      `Defaults to ${FetchTokenSalesWithPaginationDto.DEFAULT_LIMIT}, max possible value ${FetchTokenSalesWithPaginationDto.MAX_LIMIT}. ` +
+      "Note you will likely get less results than the limit, because the limit is applied before additional filtering."
+  })
+  @IsOptional()
+  @Max(FetchTokenSalesWithPaginationDto.MAX_LIMIT)
+  @Min(1)
+  @IsInt()
+  limit?: number;
+}
+
+@JSONSchema({
+  description:
+    "Response DTO for FetchTokenSwaps{ByInstanceOffered|ByInstanceWanted|OfferedBy|OfferedTo} methods. Supports pagination."
+})
+export class FetchTokenSalesWithPaginationResponse extends ChainCallDTO {
+  @IsOptional()
+  public nextPageBookmark?: string | undefined;
+
+  @ValidateNested({ each: true })
+  @Type(() => TokenSale)
+  public results: TokenSale[];
+}
+
+export class FetchTokenSaleByIdDto extends ChainCallDTO {
+  @JSONSchema({
+    description: "Token sale ID."
+  })
+  @IsNotEmpty()
+  public tokenSaleId: string;
+}
+
+@JSONSchema({
+  description:
+    "Ensure that provided tokenSaleId chain objects have proper supporting index Chain Objects. " +
+    "This method permits migration / upgrade of TokenSales written prior to the advent of new index objects " +
+    "which were added to support fine-grained querying of TokenSale by client applications. " +
+    "The provided tokenSaleIds will be looked up on chain, then their corresponding index Chain objects will be queried and written if necessary."
+})
+export class EnsureTokenSaleIndexingDto extends ChainCallDTO {
+  @ArrayNotEmpty()
+  tokenSaleIds: string[];
+}
+
+@JSONSchema({
+  description:
+    "Response DTO for a EnsureTokenSaleIndexing request. If any writes were made, they will be provided in the writes array and noOp will be true."
+})
+export class EnsureTokenSaleIndexingResponse extends ChainCallDTO {
+  @IsBoolean()
+  noOp: boolean;
+
+  @ArrayMinSize(0)
+  writes: ChainObject[];
+}
+
+// Errors
+
+export class TokenSaleDtoValidationError extends ValidationFailedError {
+  constructor(dtoName: string, errors: string[]) {
+    super(`${dtoName} validation failed: ${errors.join(". ")}`, {
+      dtoName,
+      errors
+    });
+  }
+}
+
+export class TokenSaleFailedError extends DefaultError {
```

Comment:
`DefaultError` ends up with `500` http code, probably a different base - maybe validation failed also? It has `400`

Reply:
yes, it is very experimental proof of concept. I think we do eventually want this in prod, but probably requires at least another week of work to be prod ready